server/src/main/java/org/elasticsearch/common/bytes/RecyclingBytesStreamOutput.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.bytes;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefIterator;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.BytesStream;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.ByteArray;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * An in-memory {@link StreamOutput} which first fills the given {@code byte[]} and then allocates more space from the given
+ * {@link BigArrays} if needed. The idea is that you can use this for passing data to an API that requires a single {@code byte[]} (or a
+ * {@link org.apache.lucene.util.BytesRef}) which you'd prefer to re-use if possible, avoiding excessive allocations, but which may not
+ * always be large enough.
+ */
+public class RecyclingBytesStreamOutput extends BytesStream {
+
+    private final byte[] buffer;
```

Comment:
I wonder if we need this as long as `buffer` is smaller than the page size used by `BigArrays`? Maybe see my suggested change here: https://github.com/elastic/elasticsearch/pull/61183/files#diff-80cb0d35f54671354739d96c0d8fc65aR61 
Maybe we could do an analogous thing here and just start with a `ByteArray` of capacity `PageCacheRecycler.PAGE_SIZE_IN_BYTES` that we pull from the pool and increase it's size as needed and get rid of having to allocate this array as long as it's under 16k, only allocating `buffer` if we have to materialize anything larger than `16k` in `toBytesRef` and otherwise just unwrapping the `ByteArray`? 
Then we might only have to allocate something for `buffer` once we exceed `16k` for the size of a piece of metadata but also avoid the expensive case where we allocate `size >> 3` redundantly for every write for a large value of `size` when someone's dealing with large metadata?

Reply:
Yeah ok that's not too bad actually, see ffa7726.